### PR TITLE
feat(client): warn endpoint url submission

### PR DESCRIPTION
**Pull request by ShaunSHamilton**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Feature internally discussed:
- Campers have been known to submit URLs like: `https://camper.replit.com/json`
  - This fails tests
- We should still allow endpoints, because some (me) host multiple projects on a single domain

Currently, this just adds a warning (form is **always** submit-able):
image

_I always forget: should all language translations be updated or not?_

### CODE REVIEW

File: client/i18n/locales/english/translations.json [hunk 1]
```diff
@@ -596,7 +596,8 @@
     "editor-url": "Remember to submit the Live App URL.",
     "http-url": "An unsecure (http) URL cannot be used.",
     "own-work-url": "Remember to submit your own work.",
-    "publicly-visible-url": "Remember to submit a publicly visible app URL."
+    "publicly-visible-url": "Remember to submit a publicly visible app URL.",
+    "endpoint-url": "You probably do not mean to submit a URL including an endpoint."
```
Comment (ojeytonwilliams) [inline client/i18n/locales/english/translations.json:600]:
```suggestion
    "endpoint-url": "You probably want to submit the root path i.e. https://example.com, not https://example.com/path"
```
All paths are endpoints, right? Also, examples are always nice.

File: client/src/components/formHelpers/form-validators.tsx [hunk 2]
```diff
@@ -9,6 +9,18 @@ const fCCRegex =
...
+function hasEndpoint(urlString: string): boolean {
+  try {
+    const url = new URL(urlString);
+    if (url.pathname && url.pathname !== '/') {
+      return true;
+    }
+  } catch (_e) {
+    return false;
+  }
+  return false;
```
Comment (ojeytonwilliams) [inline client/src/components/formHelpers/form-validators.tsx:21]:
```suggestion
  try {
    return new URL(urlString).pathname !== '/';
  } catch {
    return false;
  }
```
`if (condition) { return true } else { return false }` (which this boiled down to) is a bit of an anti-pattern.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: client/src/components/formHelpers/form-validators.tsx (hunk 2 above)
Comment (ShaunSHamilton) [reply client/src/components/formHelpers/form-validators.tsx:21]: 🤫

File: client/i18n/locales/english/translations.json (hunk 1 above)
Comment (ShaunSHamilton) [reply client/i18n/locales/english/translations.json:600]: Examples are always nice! Thank you.

File: client/src/components/formHelpers/form-validators.tsx [hunk 3]
```diff
@@ -9,6 +9,14 @@ const fCCRegex =
 const localhostRegex = /localhost:/;
 const httpRegex = /http(?!s|([^s]+?localhost))/;
 
+function hasEndpoint(urlString: string): boolean {
```
Comment (ojeytonwilliams) [inline client/src/components/formHelpers/form-validators.tsx:12]:
```suggestion
function isPathRoot(urlString: string): boolean {
```
Maybe? It's not super confusing as is, but it's not technically correct.

You can merge either way, but I thought I'd mention it.

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1:

File: client/i18n/locales/english/translations.json [hunk 4]
```diff
@@ -596,7 +596,8 @@
     "editor-url": "Remember to submit the Live App URL.",
     "http-url": "An unsecure (http) URL cannot be used.",
     "own-work-url": "Remember to submit your own work.",
-    "publicly-visible-url": "Remember to submit a publicly visible app URL."
+    "publicly-visible-url": "Remember to submit a publicly visible app URL.",
+    "endpoint-url": "You probably want to submit the root path i.e. https://example.com, not https://example.com/path"
```
Comment (ShaunSHamilton) [inline client/i18n/locales/english/translations.json:600]:
```suggestion
    "path-url": "You probably want to submit the root path i.e. https://example.com, not https://example.com/path"
```

File: client/src/components/formHelpers/form-fields.tsx [hunk 5]
```diff
@@ -15,7 +15,8 @@ import {
   localhostValidator,
   composeValidators,
   fCCValidator,
-  httpValidator
+  httpValidator,
+  endpointValidator
```
Comment (ShaunSHamilton) [inline client/src/components/formHelpers/form-fields.tsx:19]:
```suggestion
  pathValidator
```

File: client/src/components/formHelpers/form-fields.tsx [hunk 6]
```diff
@@ -62,7 +63,8 @@ function FormFields(props: FormFieldsProps): JSX.Element {
       name === 'githubLink' || isEditorLinkAllowed ? null : editorValidator,
       fCCValidator,
       httpValidator,
-      isLocalLinkAllowed ? null : localhostValidator
+      isLocalLinkAllowed ? null : localhostValidator,
+      endpointValidator
```
Comment (ShaunSHamilton) [inline client/src/components/formHelpers/form-fields.tsx:67]:
```suggestion
      pathValidator
```

File: client/src/components/formHelpers/form-validators.tsx [hunk 7]
```diff
@@ -23,6 +31,9 @@ export const localhostValidator = (value: string): React.ReactElement | null =>
 export const httpValidator = (value: string): React.ReactElement | null =>
   httpRegex.test(value) ? <Trans>validation.http-url</Trans> : null;
 
+export const endpointValidator = (value: string): React.ReactElement | null =>
+  hasEndpoint(value) ? <Trans>validation.endpoint-url</Trans> : null;
```
Comment (ShaunSHamilton) [inline client/src/components/formHelpers/form-validators.tsx:35]:
```suggestion
export const pathValidator = (value: string): React.ReactElement | null =>
  isPathRoot(value) ? <Trans>validation.path-url</Trans> : null;
```

**Review by raisedadead — APPROVED** [review]

**Pull request merged**